diff --git a/src/definitions.h b/src/definitions.h
index 776fd4f..14b667a 100644
--- a/src/definitions.h
+++ b/src/definitions.h
@@ -66,8 +66,12 @@
 #elif defined(HAVE_NCURSES_H)
 #include <ncurses.h>
 #else
+#if defined(__MVS__)
+#include <ncurses/ncurses.h>
+#else
 #include <curses.h>
 #endif
+#endif
 
 /* Native language support. */
 #ifdef ENABLE_NLS
diff --git a/src/text.c b/src/text.c
index 776e0d0..a921078 100644
--- a/src/text.c
+++ b/src/text.c
@@ -2119,10 +2119,13 @@ void treat(char *tempfile_name, char *theprogram, bool spelling)
 				statusline(AHEM, _("Buffer is empty"));
 			return;
 		}
-
+#if defined(__MVS__)
+                timestamp_sec = (long)fileinfo.st_mtime;
+#else
 		timestamp_sec = (long)fileinfo.st_mtim.tv_sec;
 		timestamp_nsec = (long)fileinfo.st_mtim.tv_nsec;
-	}
+#endif
+        }
 
 	/* The spell checker needs the screen, so exit from curses mode. */
 	if (spelling)
@@ -2171,8 +2174,13 @@ void treat(char *tempfile_name, char *theprogram, bool spelling)
 
 	/* When the temporary file wasn't touched, say so and leave. */
 	if (timestamp_sec > 0 && stat(tempfile_name, &fileinfo) == 0 &&
+#if defined(__MVS__)
+                                        (long)fileinfo.st_mtime == timestamp_sec) 
+#else
 					(long)fileinfo.st_mtim.tv_sec == timestamp_sec &&
-					(long)fileinfo.st_mtim.tv_nsec == timestamp_nsec) {
+					(long)fileinfo.st_mtim.tv_nsec == timestamp_nsec) 
+#endif
+        {
 		statusline(REMARK, _("Nothing changed"));
 		return;
 	}
diff --git a/src/nano.c b/src/nano.c
index 091d3ca..53c9fbb 100644
--- a/src/nano.c
+++ b/src/nano.c
@@ -857,7 +857,11 @@ bool scoop_stdin(void)
 							"type ^D or ^D^D to finish.\n"));
 
 	/* Open standard input. */
+        #if defined(__MVS__)
+	stream = fopen("/dev/fd0", "rb");
+        #else
 	stream = fopen("/dev/stdin", "rb");
+        #endif
 	if (stream == NULL) {
 		int errnumber = errno;
 
